web/src/elements/Base.ts

```diff
@@ -1,139 +1,152 @@
+    }
 
-        if (nextUITheme === UiThemeEnum.Dark) {
-            this.#darkThemeStyleSheet ||= createStyleSheetUnsafe(ThemeDark);
-            appendStyleSheet(this.#styleRoot, this.#darkThemeStyleSheet);
-            this.activeTheme = UiThemeEnum.Dark;
-        } else if (this.#darkThemeStyleSheet) {
-            removeStyleSheet(this.#styleRoot, this.#darkThemeStyleSheet);
-            this.#darkThemeStyleSheet = null;
-            this.activeTheme = UiThemeEnum.Light;
-        }
-    };
+    //#endregion
 
-    protected createRenderRoot(): HTMLElement | DocumentFragment {
-        const renderRoot = super.createRenderRoot();
-        this.#styleRoot = resolveStyleSheetParent(renderRoot);
+    //#region Properties
 
-        if (this.#customCSSStyleSheet) {
-            console.debug(`authentik/element[${this.tagName.toLowerCase()}]: Adding custom CSS`);
+    /**
+     * The resolved theme of the current element.
+     *
+     * @remarks
+     *
+     * Unlike the browser's current color scheme, this is a value that can be
+     * resolved to a specific theme, i.e. dark or light.
+     */
+    @property({
+        attribute: "theme",
+        type: String,
+        reflect: true,
+    })
+    public activeTheme: ResolvedUITheme;
+
+    //#endregion
+
+    //#region Private Properties
+
+    /**
+     * The preferred color scheme used to look up the UI theme.
+     */
+    protected readonly preferredColorScheme: CSSColorSchemeValue;
+
+    /**
+     * A custom CSS style sheet to apply to the element.
+     */
+    readonly #customCSSStyleSheet: CSSStyleSheet | null;
+
+    /**
+     * A controller to abort theme updates, such as when the element is disconnected.
+     */
+    #themeAbortController: AbortController | null = null;
+    /**
+     * The style root to which the theme is applied.
+     */
+    #styleRoot?: StyleRoot;
 
-            appendStyleSheet(this.#styleRoot, this.#customCSSStyleSheet);
+    protected set styleRoot(nextStyleRoot: StyleRoot | undefined) {
+        this.#themeAbortController?.abort();
+
+        this.#styleRoot = nextStyleRoot;
+
+        if (!nextStyleRoot) return;
+
+        if (this.#customCSSStyleSheet) {
+            setAdoptedStyleSheets(nextStyleRoot, (currentStyleSheets) => {
+                return [...currentStyleSheets, this.#customCSSStyleSheet!];
+            });
         }
 
         this.#themeAbortController = new AbortController();
 
-        if (this.#preferredColorScheme === "dark") {
-            this.#dispatchTheme(UiThemeEnum.Dark);
-        } else if (this.#preferredColorScheme === "auto") {
-            createUIThemeEffect(this.#dispatchTheme, {
+        if (this.preferredColorScheme === "dark") {
+            applyUITheme(nextStyleRoot, UiThemeEnum.Dark);
+
+            this.activeTheme = UiThemeEnum.Dark;
+        } else if (this.preferredColorScheme === "auto") {
+            createUIThemeEffect((nextUITheme) => applyUITheme(nextStyleRoot, nextUITheme), {
```

Comment:
Small revision to split off the logic of applying a theme into a function that Storybook and WebDriver's test runner can use.